test/cpp/end2end/connection_delay_injector.h

```diff
@@ -67,26 +68,28 @@ class ConnectionAttemptInjector {
    public:
     QueuedAttempt(grpc_closure* closure, grpc_endpoint** ep,
                   grpc_pollset_set* interested_parties,
-                  const grpc_channel_args* channel_args,
+                  TcpGenericOptions&& options,
                   const grpc_resolved_address* addr,
                   grpc_core::Timestamp deadline)
         : closure_(closure),
           endpoint_(ep),
           interested_parties_(interested_parties),
-          channel_args_(grpc_channel_args_copy(channel_args)),
+          options_(std::move(options)),
           deadline_(deadline) {
       memcpy(&address_, addr, sizeof(address_));
     }
 
     ~QueuedAttempt() {
       GPR_ASSERT(closure_ == nullptr);
-      grpc_channel_args_destroy(channel_args_);
+      grpc_tcp_generic_options_destroy(&options_);
     }
 
     // Caller must invoke this from a thread with an ExecCtx.
     void Resume() {
       GPR_ASSERT(closure_ != nullptr);
-      AttemptConnection(closure_, endpoint_, interested_parties_, channel_args_,
+      auto config = grpc_event_engine::experimental::CreateEndpointConfig(
+          TcpOptionsIntoChannelArgs(options_));
```

Comment:
I agree, we shouldn't be using `TcpGenericOptions` in this code.  Instead, I suggest having this code down-cast from `EndpointConfig` to `ChannelArgsEndpointConfig`, which should be made copyable, and we can then store that internally when we delay a connection attempt.

Craig, you're right, we're definitely going to need to make some changes to this library as part of the EE migration, since this code currently reaches into iomgr internals to work.  In the long run, I was thinking that we would basically make this a wrapper EE impl that wraps around a real EE impl so that it can intercept the relevant API calls.  You're right that we could also consider making it a handshaker, but I suspect that will be a little harder to implement, and it would be slightly less general-purpose, because there may be use-cases in which we create a TCP connection but don't use handshakers (although to be fair, we are not actually using this library for any such case right now).

But yeah, let's definitely talk about how to deal with this during the migration.  The interim phases of the migration may be a little challenging here.

Reply:
I implemented mark's suggestion in the lates commit.